### ISSUE TITLE
Upgrade to maven 3.9.0, resolver 1.9.4

### ISSUE THREAD

**Issue by ponziani** [ops4j/org.ops4j.pax.url#417]
In order to be able to read snapshots from a separate local maven-repository from Apache Karaf, the Maven version should be upgraded.

https://issues.apache.org/jira/browse/MNG-6270
https://issues.apache.org/jira/browse/KARAF-7667

I tried building the project myself to create a PR, but unfortunately I faced several failing tests.

**Comment by grgrzybek** [comment]
it's not that easy. Maven Resolver 1.9 is quite new and requires few tricks (it deprecates service locator). I used it in Camel, so it's possible, but not just by bumping the version. Also Maven version itself is not that important here to leverage MNG-6270.

However, Karaf uses pax-url-aether where you can configure multiple repositories simply by adding them to `org.ops4j.pax.url.mvn.defaultRepositories` property of `org.ops4j.pax.url.mvn` PID.

**Comment by grgrzybek** [comment]
For karaf-maven-plugin, unfortunately `org.apache.karaf.profile.assembly.Builder#createMavenResolver()` used by K-M-P, is setting only 3 properties which are taken from Maven plugin `<configuration>`:
https://github.com/apache/karaf/blob/karaf-4.4.3/profile/src/main/java/org/apache/karaf/profile/assembly/Builder.java#L1473-L1486

However you can create KARAF Jira issue with an improvement suggestion.

**Comment by cstamas** [comment]
Just FTR, the `main` should be buildable, while `master` will have failing tests. Unsure why both exists, but this is what it is.

**Comment by grgrzybek** [comment]
`master` was simply left behind and `main` was used after Pax URL 2.6.11 release.

**Comment by grgrzybek** [comment]
From https://lists.apache.org/thread/j41t2gr1nl6w886nqhdyzfk5p5c27x36:

> I raised a ticket for Karaf itself, but forgot I meant to raised it on
pax-url - see my blog post on it:
>
>
> https://www.theoryinpractice.net/post/708585430266855424/healing-the-poisoned-repository
>
> Mark

**Comment by grgrzybek** [comment]
@cstamas hi, I can finally get back to this one (and #423). I'm more inclined to use apache/maven-resolver#319 (the new `org.eclipse.aether.supplier.RepositorySystemSupplier` class) instead of https://github.com/maveniverse/mima. Any final recommendation?

**Comment by grgrzybek** [comment]
See:
* https://maven.apache.org/guides/mini/guide-http-settings.html
* https://maven.apache.org/guides/mini/guide-resolver-transport.html
* https://maven.apache.org/resolver/configuration.html

> Historically, since Maven 2.x, the only transport used by Maven was Maven Wagon. Since the introduction of the Resolver in Maven 3.x, it did “wrap” it in the maven-resolver-transport-wagon module, and continued to use it. Still, Wagon is deeply integrated with the old Plexus DI container and many, if not all, of its configuration are bound to be set in Plexus XML, which is not type safe, nor validated, nor future-proof.
>
> Starting with Maven 3.9.0 release, the default transport (the default transport used by Maven Resolver) changed from ancient Wagon to modern maven-resolver-transport-http aka native HTTP transport.

**Comment by cstamas** [comment]
If you go with 'bare' resolver, you need to duplicate all the logic that exists "between" resolver and "maven", like: settings.xml discovery, parsing, decryption, populating that (reposes, properties) somehow etc... MIMA gives you all this out of the box. I'd recommend MIMA, but as you wish :smile:

**Comment by grgrzybek** [comment]
Thanks for reminding. Pax URL needs (and already has) quite a lot of control and we manage (and decrypt) settings anyway using OSGi Configuration Admin. So I think sticking to `org.apache.maven.resolver` and `org.apache.maven` is better in Pax URL currently and in OSGi in general.

Previous approach with `org.apache.maven.repository.internal.MavenRepositorySystemUtils#newServiceLocator()` needed to configure everything that's needed and with `org.eclipse.aether.supplier.RepositorySystemSupplier` I can simply change what's needed, so it's much easier and clear.

Thanks for feedback!

**Comment by grgrzybek** [comment]
The biggest change will be to switch from wagon to maven-resolver-transport-http. With wagon, Pax URL configured its own HttpClient and here it _may_ not be possible.
So getting rid of wagon may not be possible...

**Comment by cstamas** [comment]
Cool, ok. And a gentle reminder:
* resolver 1.x needs "proper shutdown" for RepositorySystem, don't forget that

Also, as you are about to do changes, prepare for these coming in Resolver 2.x:
* session in 2.x is "closeable" and if you opt to use HTTP/2 transport, you MUST close it (to prevent leaks) https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/RepositorySystemSession.java#L65
* added session supplier as well https://github.com/apache/maven-resolver/tree/master/maven-resolver-supplier/src/main/java/org/eclipse/aether/supplier

So just prepare for future:
* have place to close (since 2.x) or shutdown (in 1.x) the RepositorySystem
* have place and provide proper "boxing" for closeable session

**Comment by cstamas** [comment]
Also, if you want to "take a peek" there is 2.0.0-alpha-1 already on central (does not have onSessionClose), but better wait for 2.0.0-alpha-2 am doing in a day or two that will include closeable sessions as well...

**Comment by cstamas** [comment]
Alpha-2 will be those closed issues that are not in alpha-1 :smile: 
https://issues.apache.org/jira/issues/?jql=project%20%3D%20MRESOLVER%20AND%20fixVersion%20%3D%202.0.0

**Comment by cstamas** [comment]
And more question and will stop: would it help if mima (I guess "context" and "standalone-static" runtime, as that is what you need) become OSGi bundles (have proper metadata)? :wink:

**Comment by grgrzybek** [comment]
Great - Pax URL would be great opportunity not only to move pax-url-war to jakarta.servlet API, but also Maven Resolver 2.

> resolver 1.x needs "proper shutdown" for RepositorySystem

oops - I forgot to call `org.eclipse.aether.internal.impl.DefaultRepositorySystem#shutdown()`... But looks like it calls `org.eclipse.aether.internal.impl.DefaultRepositorySystemLifecycle#systemEnded()` and iterates over `DefaultRepositorySystemLifecycle#onSystemEndedHandlers` - are these added implicitly? I never added such handler in Pax URL. But anyway - I'll remember!

> added session supplier as well https://github.com/apache/maven-resolver/tree/master/maven-resolver-supplier/src/main/java/org/eclipse/aether/supplier

Didn't we reject https://issues.apache.org/jira/browse/MRESOLVER-394 before?

**Comment by grgrzybek** [comment]
> And more question and will stop: would it help if mima (I guess "context" and "standalone-static" runtime, as that is what you need) become OSGi bundles (have proper metadata)? 😉

I would have to check...

Pax URL includes (via `Private-Package`) all maven stuff anyway, so it wouldn't be a big difference. Why embed it you may ask? Pax URL is very low-level bundle and is used almost immediately (in terms if launching OSGI runtime) in Karaf at very low start-level value.
This is important, because all _later_ bundles are installed via `mvn:` protocol. If we (Karaf) try to do it in proper OSGi way, instead of this:
```
karaf@root()> la -l
START LEVEL 100 , List Threshold: 0
ID │ State     │ Lvl │ Version            │ Location
───┼───────────┼─────┼────────────────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 0 │ Active    │   0 │ 7.0.5              │ System Bundle
 1 │ Resolved  │   1 │ 4.4.4              │ mvn:org.apache.karaf.features/org.apache.karaf.features.extension/4.4.4
 2 │ Active    │   5 │ 4.4.4              │ mvn:org.apache.karaf.services/org.apache.karaf.services.eventadmin/4.4.4
 3 │ Active    │   5 │ 1.2.4              │ mvn:org.apache.felix/org.apache.felix.metatype/1.2.4
 4 │ Active    │   5 │ 2.6.14             │ mvn:org.ops4j.pax.url/pax-url-aether/2.6.14
...
```

We'd get many more bundles before pax-url-aether (all Maven, optionally Wagon, HttpClient, Resolver, Plexus, ...) touching (refreshing) these bundles would lead to refresh of many many bundles. That's why repackaged libraries inside Pax URL is better (from maintenance, deployment and administration perspective) than purist OSGi approach...

**Comment by cstamas** [comment]
Just wanted to say, that "standalone-static-uber" IS repackaged to have it all and I plan to keep it that way (and publish various combos of uber runtimes, as needed).

**Comment by cstamas** [comment]
> Didn't we reject https://issues.apache.org/jira/browse/MRESOLVER-394 before?

AFAIR, we did not reject, but we areeed o.a.maven:maven-resolver-provider MavenRepoSystemUtils is "enough". But, v2 changes this, look at related changes here:
https://github.com/apache/maven/pull/1301

**Comment by grgrzybek** [comment]
> Just wanted to say, that "standalone-static-uber" IS repackaged to have it all and I plan to keep it that way (and publish various combos of uber runtimes, as needed).

I'll check it then ;)

**Comment by cstamas** [comment]
In essence, v2 uses closeable session, but it was done in a way that:
* "resolver embedding" apps (like Maven) needs to change their session handling (would work with "old way" as well, but w/ leaks)
* "resolver using" code (like Mojos) does NOT need change, they are oblivious that session is closeable, nor should they EVER close it

**Comment by cstamas** [comment]
> Great - Pax URL 3 would be great opportunity not only to move pax-url-war to jakarta.servlet API, but also Maven Resolver 2.
> 
> > resolver 1.x needs "proper shutdown" for RepositorySystem
> 
> oops - I forgot to call `org.eclipse.aether.internal.impl.DefaultRepositorySystem#shutdown()`... But looks like it calls `org.eclipse.aether.internal.impl.DefaultRepositorySystemLifecycle#systemEnded()` and iterates over `DefaultRepositorySystemLifecycle#onSystemEndedHandlers` - are these added implicitly? I never added such handler in Pax URL. But anyway - I'll remember!

The system internally uses those hooks, but your code could use it as well (RepoSystem iface has method to add hooks).

**Comment by grgrzybek** [comment]
Just for the record.

Since maven-resolver 1.8, enhanced version of `org.eclipse.aether.repository.LocalRepositoryManager` can handle [_split repositories_](https://maven.apache.org/resolver/local-repository.html#split-local-repository).

`org.eclipse.aether.internal.impl.LocalPathPrefixComposer` uses several settings which may _inject_ subdirectories between `~/.m2/repository` and actual path (starting from groupId) of the artifact.

* `-Daether.enhancedLocalRepository.split=true/false` turns on or off splitting in the first place. So without this option, no other options are effective. With `split=true` local repository is split into cached (remote) and installed (local) repositories:
  * `-Daether.enhancedLocalRepository.remotePrefix=cached` is a subdirectory for remote artifacts/metadata
  * `-Daether.enhancedLocalRepository.localPrefix=installed` is a subdirectory for local artifacts/metadata
* `-Daether.enhancedLocalRepository.splitLocal=true/false` (only if `split=true` as well) turns on or off splitting for local artifacts by SNAPSHOT and non-SNAPSHOT versions:
  * `-Daether.enhancedLocalRepository.releasesPrefix=releases` is a subdirectory for non-SNAPSHOT artifacts
  * `-Daether.enhancedLocalRepository.snapshotsPrefix=snapshots` is a subdirectory for SNAPSHOT artifacts
* `-Daether.enhancedLocalRepository.splitRemote=true/false` (only if `split=true` as well) turns on or off splitting for remote artifacts by SNAPSHOT and non-SNAPSHOT versions using the same (`releasesPrefix`, `snapshotsPrefix`) as for local artifacts
* `-Daether.enhancedLocalRepository.splitRemoteRepository=true/false` may additionally create a subdirectory for cached remote artifacts based on remote repository (or mirror?) ID
* `-Daether.enhancedLocalRepository.splitRemoteRepositoryLast=true/false` chooses between adding repo-id based subdirectory after (true) or before (false) snapshots/releses subdirectory part

When enabling all splits, we may end with something like:
```
$ pwd
~/.m2/repository

$ tree -L 4
.
├── cached
│   └── central
│       └── releases
│           ├── com
│           ├── commons-codec
│           ├── commons-io
│           ├── io
│           ├── jakarta
│           ├── net
│           └── org
└── installed
    └── releases
        └── com
            └── example
```

**Comment by ponziani** [comment]
Hi @grgrzybek,

I'm really sorry, but the current situation is not clear to me. Is there a possibility for me to use the enhancedLocalRepository.split-property with karaf-maven-plugin 4.4.6? If so, how ?

**Comment by grgrzybek** [comment]
@ponziani no :( the fix is in the new `url-3.0.x` branch and I didn't release 3.0.0 yet (I didn't even polish it fully)...

Also 3.0.0 will contain #426 fix where pax-url-war moves from `javax.servlet.*` to `jakarta.servlet.*`...
